src/Elastic.Apm.Azure.Functions/ElasticApmMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Apm.Api;
+using Elastic.Apm.Cloud;
+using Elastic.Apm.Helpers;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Http;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using TraceContext = Elastic.Apm.DistributedTracing.TraceContext;
+
+namespace Elastic.Apm.Azure.Functions;
+
+public class ElasticApmMiddleware : IFunctionsWorkerMiddleware
+{
+	private static readonly IApmLogger Logger;
+	private static readonly string FaasIdPrefix;
+	private static int ColdStart = 1;
+
+	static ElasticApmMiddleware()
+	{
+		Logger = Agent.Instance.Logger.Scoped(nameof(ElasticApmMiddleware));
+		var metaData = new AzureFunctionsMetadataProvider(Logger,
+			new EnvironmentVariables(Logger).GetEnvironmentVariables()).GetMetadata();
+		FaasIdPrefix =
+			$"/subscriptions/{metaData?.Account?.Id}/resourceGroups/{metaData?.Project}/providers/Microsoft.Web/sites/{metaData?.Instance}/functions/";
+		Logger.Trace()?.Log($"FaasIdPrefix: {FaasIdPrefix}");
+	}
+
+	private static bool IsColdStart() => Interlocked.Exchange(ref ColdStart, 0) == 1;
+
+	public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+	{
+		try
+		{
+			Logger.Trace()?.Log($"{nameof(Invoke)} - {context.FunctionDefinition.Name}");
+
+			var data = GetTriggerSpecificData(context);
+			await Agent.Tracer.CaptureTransaction(data.Name, ApiConstants.TypeRequest, async t =>
```

Comment:
Did you test what happens when `await next(context);`  throws?

If `await next(context);` within this block throws an exceptions, then the `CaptureTransaction` will deal with it (end the transaction properly, capture the exception, etc..), but we have 2 lines after `await next(context);` which set `t.Outcome` and `t.Result` - those will not run if `await next(context);` throws.

I haven't tested this and I don't know if those 2 lines would otherwise do something relevant in that case (I'd expect to see HTTP500 on the `context`, but maybe that's not the case).

If this is really an issue, we could just do try-catch-finally here and start the transaction by calling `Agent.Tracer.StartTransaction` and manually end it, plus run the necessary lines to set `Outcome` and `Result` in `finally`.

Reply:
I improved and refactored that code. PTAL
(see also my [comment](https://github.com/elastic/apm-agent-dotnet/pull/1891#discussion_r1026156310) above)